src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -433,50 +452,70 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
         return CompletableFuture.completedFuture(0);
     }
 
-    private synchronized void spoolMessage()  {
-        if (spoolingFuture.get() == null || spoolingFuture.get().isCancelled()) {
-            spoolingFuture.set(ses.scheduleWithFixedDelay(this::spoolTask, 0, 5, TimeUnit.SECONDS));
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    protected CompletableFuture<Integer> publishSingleSpoolerMessage() throws InterruptedException {
+        final long id = spool.popId();
+        PublishRequest request = spool.getMessageById(id);
+        if (request == null) {
+            return null;
+        }
```

Comment:
Can this actually happen in any normal condition? When would it happen? Regardless a log message is warranted.

Reply:
Also I would prefer we don't return null value unless null is a legit value in certain use case, which doesn't sound like it here.

Reply:
It is backed by a Map, so it returns null if the ID can't be found in the map.

Reply:
but why would the id not be found in the map if it's popped off from the queue? Does it get removed/updated? Is there any operation not synchronized? Sounds like only a bug can cause the situation, right?

Reply:
Seems that's the case, yes. @awszztt?

Reply:
When the spool fills up, QoS 0 messages can get dropped from the backing store, but the ID isn't removed from the queue. Really the spool should hide that

Reply:
@jbutler Yes, that is the case.

Reply:
Why can't the id be removed as well? This doesn't sound ideal.

Reply:
The id is stored in LinkedBlockingDeque and id of the message with QoS could be at anywhere in the queue. Wondering it is worth to do the extra operation to remove this id from the queue.

Reply:
Okay. Then why do you even to return null to the caller? Why not just pop and discard those ids until you reach one that exists in the map.  I would also add a debug logline as well.

Reply:
yes, I will make some change to achieve this.